Update opacity input slightly (add table reader option).

RyanWollaeger opened the pull request:
## Background

* This change set (hopefully) follows @pdmullen's change to Jaybenne opacity input options.
* With a Debug build, models asserted incorrect units prior to the change to non-cgs model opacity variants.

## Description of Changes

+ Add parsed tablular opacity option to input.
+ Update strat.hpp problem generator to see `do_radiation`, and add @pdmullen patch for DDMC B.C. data avoidance.

## Checklist

- [ ] New features are documented
- [ ] Tests added for bug fixes and new features
- [ ] (@lanl.gov employees) Update copyright on changed files

RyanWollaeger committed:
Update opacity input slightly.

+ Add parsed tablular opacity option to input.
+ Change model variants to non-cgs, for consistency with means.

Note: with a Debug build, models asserted incorrect units prior
to the change to non-cgs model opacity variants.

RyanWollaeger committed:
Revert model instantiations to CGS.

RyanWollaeger committed:
Apply Patrick Mullen's patch to strat.hpp to get problem working with DDMC.
(Thank you to Patrick!)

RyanWollaeger committed:
Initialize jaybenne particles if do_radiation in strat.hpp.

RyanWollaeger committed:
Update jaybenne module to bring in small particle streaming patch.

adamdempsey90 commented on src/utils/opacity/opacity.hpp:
```diff
@@ -22,9 +22,10 @@
+    singularity::photons::NonCGSUnits<singularity::photons::EPBremss>>;
 
 // Reduced scattering variant for this codebase
 using Scattering = singularity::photons::impl::S_Variant<singularity::photons::GrayS,
```
Does the Scattering opacity not need the same `NonCGSUnits` wrapper?

pdmullen replied:
Actually now I am confused too... 

I thought that `Opacity` and `Scattering` did not need the `NonCGSUnits` wrapper... and only `MeanOpacity` and `MeanScattering` needed them.  (at one point I had them on both, IIRC, and things were broken).

RyanWollaeger replied:
Indeed they don't - this `Opacity` variant is actually not getting used now (but I left it since it could get used in the future if MG opacity is supported in the moment solver) - instead in the if-check block where we set the `model`, the `model` is scoped in the if-blocks, and gets type-resolved with `auto`. This is more similar to Jaybenne.

RyanWollaeger replied:
The `NonCGSUnitsS` wrapper is now applied to the scattering variant, and a similar scoping of the CGS scattering `smodel` has been applied in `gas.cpp`, in the last commit.

adamdempsey90 commented on src/pgen/strat.hpp:
```diff
@@ -360,65 +372,68 @@ inline void ShearInnerX2(std::shared_ptr<MeshBlockData<Real>> &mbd, bool coarse)
-            v(0, dust::prim::velocity(VI(n, 1)), k, j, i) = vx2d;
-            v(0, dust::prim::velocity(VI(n, 2)), k, j, i) = vx3d;
-            v(0, dust::prim::density(n), k, j, i) = densd;
+  if (v.GetMaxNumberOfVars() > 0) {
```
I suppose we could lower the diffs and extra indentations on all of these by just returning if `v.GetMaxNumberOfVars() == 0`

RyanWollaeger replied:
Good idea - will do.

RyanWollaeger replied:
Made this change in the last commit - thanks!

RyanWollaeger committed:
Update submodules and simplify strat.hpp change.

+ Update jaybenne and singularity submodules to fix Debug builds.
+ Return on 0 max number of variables in strat.hpp B.C.s.

RyanWollaeger committed:
Apply similar model-scoping to scattering/mean scattering.

adamdempsey90 reviewed (approved):
I'd like to get this merged asap so I can finish moments. Are we ready to merge?

RyanWollaeger commented:
> I'd like to get this merged asap so I can finish moments. Are we ready to merge?

No objection here - no further changes planned - @pdmullen?

pdmullen committed:
Merge branch 'develop' into tabl_opac
